drv/stm32h7-eth/src/ring.rs

```diff
@@ -0,0 +1,429 @@
+    }
+
+    /// Returns the count of entries in the descriptor ring / buffers in the
+    /// pool.
+    pub fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    pub fn is_next_free(&self) -> bool {
+        let d = &self.storage[self.next.get()];
+        // Check whether the hardware has released this.
+        let tdes3 = d.tdes[3].load(Ordering::Acquire);
+        let own = tdes3 & (1 << TDES3_OWN_BIT) != 0;
+        !own
+    }
+
+    /// Attempts to grab the next unused TX buffer in the ring and deposit a
+    /// packet into it.
+    ///
+    /// If the next buffer in the ring is not holding a pending packet, this
+    /// will borrow it and call `body` with its address. `body` returns the
+    /// number of bytes in the packet; this routine will then set up a pending
+    /// transmit descriptor for that prefix of the buffer, and return
+    /// `Some(len)`.
+    ///
+    /// If the next buffer in the ring is pending, that means we have run out of
+    /// TX ring slots. In that case, this will return `None` without invoking
+    /// `body`.
+    ///
+    /// # Panics
+    ///
+    /// If a buffer is available, and we call `body`, and `body` returns a valid
+    /// length larger than `BUFSZ`. Because that's obviously wrong.
+    pub fn try_with_next<R, F: FnOnce(&mut [u8]) -> R>(
+        &self,
+        len: usize,
+        body: F,
+    ) -> Result<R, F> {
+        let d = &self.storage[self.next.get()];
+        // Check whether the hardware has released this.
+        let tdes3 = d.tdes[3].load(Ordering::Acquire);
+        let own = tdes3 & (1 << TDES3_OWN_BIT) != 0;
+        if own {
+            Err(body)
+        } else {
+            // Descriptor is free. Since we keep the descriptors paired with
+            // their corresponding buffers, and only lend out the buffers
+            // temporarily (like we're about to do), this means the buffer is
+            // also free.
+            let buffer = self.buffers[self.next.get()].0.get();
+            // Safety: we're dereferencing a raw *mut to get a &mut into the
+            // buffer. We must ensure that the pointer is valid (which we can
+            // tell trivially from the fact that we got it from an UnsafeCell)
+            // and that there is no aliasing. We can say there's no aliasing
+            // because the descriptor is free (above), so we're about to produce
+            // the sole reference to it, which won't outlive this block.
+            let buffer = unsafe { &mut *buffer };
+            let buffer = &mut buffer[..len];
+
+            let result = body(buffer);
+
+            // Program the descriptor to represent the packet. We program
+            // carefully to ensure that the memory accesses happen in the right
+            // order: the entire descriptor must be written before the OWN bit
+            // is set in TDES3 using a RELEASE store.
+            d.tdes[0].store(buffer.as_ptr() as u32, Ordering::Relaxed);
```

Comment:
Could we make the `struct TxDesc` have four members with meaningful names, instead of this array?

Reply:
They're named `TDES0` through `TDES3` in the reference manual. I had fields named that at one point and then switched to the array. There aren't obviously better names available imo.